src/asynqp/routing.py

```diff
@@ -26,13 +26,31 @@ def dispatch(self, frame):
         writer = self.queue_writers[frame.channel_id]
         writer.enqueue(frame)
 
+    def dispatch_all(self, frame):
+        for writer in self.queue_writers.values():
+            writer.enqueue(frame)
+
 
 class Synchroniser(object):
+    _blocking_methods = set((spec.BasicCancelOK,  # Consumer.cancel
+                             spec.ChannelCloseOK,  # Channel.close
+                             spec.ConnectionCloseOK))  # Connection.close
+
     def __init__(self):
         self._futures = OrderedManyToManyMap()
+        self.connection_closed = False
 
     def await(self, *expected_methods):
         fut = asyncio.Future()
+
+        if self.connection_closed:
+            for method in expected_methods:
+                if method in self._blocking_methods and not fut.done():
+                    fut.set_result(None)
+            if not fut.done():
```

Comment:
We have three methods in self._blocking_methods. The first time through the loop, fut.done() will be False, but if for some reason you sent more than one of them in *expected_methods, the future would be set twice raising an InvalidStateError.

If none of the _blocking_methods are included, the loop proceeds and sets an exception to kill the waiting task.